python/BioSimSpace/_SireWrappers/_system.py

```diff
@@ -1702,6 +1758,47 @@ def _set_water_topology(self, format, property_map={}):
                 self._sire_object = _SireIO.setGromacsWater(
                         self._sire_object, water_model, property_map)
 
+    def _set_atom_index_tally(self):
+        """Internal helper function to create a dictionary mapping molecule
+           numbers to the cumulative total of atoms in the system.
+        """
+        # Only compute the atom index mapping if it hasn't already
+        # been created.
+        if len(self._atom_index_tally) == 0:
+            # Loop over all molecules in the system and keep track
+            # of the cumulative number of atoms.
+            num_atoms = 0
+            for num in self._sire_object.molNums():
+                self._atom_index_tally[num] = num_atoms
+                num_atoms += self._sire_object.molecule(num).nAtoms()
```

Comment:
Should these be checked against the `Sire` `molNums` or the `BioSimSpace` numbers which follow the order of the molecules? I imagine they are supposed to always be the same but in case they are not, this will result in unexpected behaviour?

Reply:
It should always be the same but I'll check.

Reply:
Yes, this holds in my checks, i.e. performing various system modifications and making sure the `MolNums` are reconstructed in the correct order, depending on how the System was created. This did raise one small bug, where adding a `Molecules` object to a `Molecule` resulted in the molecule being added to the _end_ of the existing `Molecules` container, rather than a new container being returned with the `Molecule` at the start. This didn't affect any indexing operations on the resulting container, only that the items in the container were in an unexpected order to start with.